SmartTutor/ContentModel/LearningObjects/ChallengeModel/FulfillmentStrategy/MetricChecker/MetricRangeRule.cs

```diff
@@ -12,10 +12,389 @@ public class MetricRangeRule
         public double FromValue { get; internal set; }
         public double ToValue { get; internal set; }
 
-        internal bool MetricMeetsRequirements(Dictionary<CaDETMetric, double> metrics)
+        public bool MetricMeetsRequirements(Dictionary<CaDETMetric, double> metrics)
         {
             var metricValue = metrics[(CaDETMetric)Enum.Parse(typeof(CaDETMetric), MetricName, true)];
             return (FromValue <= metricValue && metricValue <= ToValue);
         }
+
+        public ChallengeHint GetHintForIncompleteClass()
```

Comment:
It seems I didn't communicate well enough the intent of the ChallengeHint Content field. Basically, we would want this to be some additional information that is not present in the learning object, which the guy creating the challenge would manually populate in the database.

I feel bad saying this, but we really don't need the code below. First, it is hardcoded with English words. Second, most trainees won't know what the more elaborate metrics mean so they won't be sure how to comply with the advice. Third, advice such as "decrease the cohesion in the class" is not something we would ever recommend, while advice such as "increase the number of protected fields" is not a metric rule we would define. Fourth, every time we would add a metric we would need to modify these rules in the Smart Tutor even though our existing MetricRangeRules don't require the new metric.

The takeaway is - each MetricRangeRule has zero or one ChallengeHint associated with them. The strategy can then see if the rule passes, and if not see if the rule has a hint it can offer. This will be similar for the NamingRules.

If this requires further clarification we can talk over Teams.